File: backend/functions/src/cleanup_history.ts
import * as admin from "firebase-admin";
import * as logger from "firebase-functions/logger";

export interface HistoryCleanupResult {
  success: boolean;
  totalDeleted: number;
  batchesProcessed: number;
  durationMs: number;
}

/**
 * Core handler to process batch deletion of expired task history documents across Firestore.
 * Queries db.collectionGroup("history") for documents where expiresAt <= now.
 *
 * @param db - Firestore database instance.
 * @param now - Reference timestamp threshold for expiration (defaults to admin.firestore.Timestamp.now()).
 * @param batchLimit - Number of documents to delete per batch (max 500 per Firebase guidelines).
 * @param maxBatches - Safe limit on number of batch iterations per execution to prevent timeouts.
 */
export async function processHistoryCleanup(
  db: admin.firestore.Firestore,
  now: admin.firestore.Timestamp = admin.firestore.Timestamp.now(),
  batchLimit: number = 500,
  maxBatches: number = 20
): Promise<HistoryCleanupResult> {
  const startTime = Date.now();
  let totalDeleted = 0;
  let batchesProcessed = 0;

  try {
    while (batchesProcessed < maxBatches) {
      const snapshot = await db
        .collectionGroup("history")
        .where("expiresAt", "<=", now)
        .limit(batchLimit)
        .get();

      if (snapshot.empty) {
        break;
      }

      const batch = db.batch();
      for (const doc of snapshot.docs) {
        batch.delete(doc.ref);
      }

      await batch.commit();
      totalDeleted += snapshot.size;
      batchesProcessed++;

      if (snapshot.size < batchLimit) {
        break;
      }
    }

    const durationMs = Date.now() - startTime;
    logger.info(
      `History cleanup completed successfully: deleted ${totalDeleted} documents across ${batchesProcessed} batches in ${durationMs}ms`
    );

    return {
      success: true,
      totalDeleted,
      batchesProcessed,
      durationMs,
    };
  } catch (error) {
    const durationMs = Date.now() - startTime;
    logger.error("Error during history cleanup processing:", error);
    throw error;
  }
}

File: backend/functions/lib/src/handlers/cleanup_history.dart
import 'package:meta/meta.dart';
import '../interop/firebase_admin.dart';
import '../interop/node_interop.dart';

@immutable
class HistoryCleanupResult {
  final bool success;
  final int totalDeleted;
  final int batchesProcessed;
  final int durationMs;

  const HistoryCleanupResult({
    required this.success,
    required this.totalDeleted,
    required this.batchesProcessed,
    required this.durationMs,
  });

  Map<String, dynamic> toJson() {
    return {
      'success': success,
      'totalDeleted': totalDeleted,
      'batchesProcessed': batchesProcessed,
      'durationMs': durationMs,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is HistoryCleanupResult &&
        other.success == success &&
        other.totalDeleted == totalDeleted &&
        other.batchesProcessed == batchesProcessed &&
        other.durationMs == durationMs;
  }

  @override
  int get hashCode => Object.hash(
        success,
        totalDeleted,
        batchesProcessed,
        durationMs,
      );
}

/// Core handler to process batch deletion of expired task history documents across Firestore.
/// Queries db.collectionGroup("history") for documents where expiresAt <= now.
///
/// [db]: Firestore database instance.
/// [now]: Reference timestamp threshold for expiration.
/// [batchLimit]: Number of documents to delete per batch (max 500 per Firebase guidelines).
/// [maxBatches]: Safe limit on number of batch iterations per execution to prevent timeouts.
Future<HistoryCleanupResult> processHistoryCleanup(
  FirestoreDatabase db, [
  dynamic now,
  int batchLimit = 500,
  int maxBatches = 20,
]) async {
  final threshold = now ?? DateTime.now().toUtc();
  final startTime = DateTime.now().millisecondsSinceEpoch;
  var totalDeleted = 0;
  var batchesProcessed = 0;

  try {
    while (batchesProcessed < maxBatches) {
      final snapshot = await db
          .collectionGroup('history')
          .where('expiresAt', '<=', threshold)
          .limit(batchLimit)
          .get();

      if (snapshot.empty) {
        break;
      }

      final batch = db.batch();
      for (final doc in snapshot.docs) {
        batch.delete(doc.ref);
      }

      await batch.commit();
      totalDeleted += snapshot.size;
      batchesProcessed++;

      if (snapshot.size < batchLimit) {
        break;
      }
    }

    final durationMs = DateTime.now().millisecondsSinceEpoch - startTime;
    logInfo(
      'History cleanup completed successfully: deleted $totalDeleted documents across $batchesProcessed batches in ${durationMs}ms',
    );

    return HistoryCleanupResult(
      success: true,
      totalDeleted: totalDeleted,
      batchesProcessed: batchesProcessed,
      durationMs: durationMs,
    );
  } catch (error) {
    final durationMs = DateTime.now().millisecondsSinceEpoch - startTime;
    logError(
      'Error during history cleanup processing after ${durationMs}ms:',
      error,
    );
    rethrow;
  }
}
